### PR TITLE
Documentation clarification and more info on using Batch

### PR DESCRIPTION
**Pull request by philvarner**
Several additions to the docs, integrating what I learned from the Earth Search Sentinel-2 ingest work.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,26 +4,29 @@
...
+STAC API catalog (e.g.,
+[stac-server](https://github.com/stac-utils/stac-server)).
 
 Cirrus workflows range from the simple publishing of unmodified input items to
 the complex transformation of input Items and generation of wholly-new output
 Items. The current state of CPP processing is tracked in a state database
 to prevent duplicate processing and allow for a user to follow the state of any
 input through the pipeline.
 
-![](docs/images/highlevel.png)
```
Comment (philvarner) [inline README.md:25]: this image link was broken, so I replaced it with the arch overview and put it near the top.

File: src/cirrus/docs/10_intro.rst [hunk 2]
```diff
@@ -29,35 +30,37 @@ Concepts
...
-  processing pipeline to transform a given input into one or more output STAC
-  items
-
-An additional component type is that of a :doc:`Function
-<components/functions>`, though they are less commonly extended by end users.
-
+<60_components>`. There are four types of Components:
+
+* :doc:`Feeders <components/feeders>` accept arbitrary input, and output a
+  Cirrus Process Payload, which is enqueue for processing. These can be used to initiate
```
Comment (stark525) [inline src/cirrus/docs/10_intro.rst:52]: enqueue -> enqueued?

File: src/cirrus/docs/10_intro.rst [hunk 3]
```diff
@@ -83,34 +86,56 @@ Relationship with stac-server
...
 
 .. _stac-server: https://github.com/stac-utils/stac-server
 
 
-Example use cases
------------------
+Example Earth Search use case
+----------------
+
+One prominent use case of Cirrus is as the processing pipeline for `Earth Search`_.
```
Comment (stark525) [inline src/cirrus/docs/10_intro.rst:98]: Is it worth digging slightly deeper into how these map to cirrus concepts? maybe even a diagram of the specific Landsat workflow with USGS's SNS topic? (this can be a separate issue / improvement later)

Comment (philvarner) [reply src/cirrus/docs/10_intro.rst:98]: I think it would be -- I think that's one of the overall problems with these docs -- they do a good job of describing everything, but I don't they give a good picture of the dynamics of the system and how an individual pipeline would be implement and flow end-to-end. I think it's later work.

File: src/cirrus/docs/components/tasks/batch.rst [hunk 4]
```diff
@@ -416,11 +416,78 @@ CPUs each job requires.
...
 *************************
 
-.. TODO
-
+Using Spot instances instead of On-Demand can result in a significant cost savings, with only a
+small effect on throughput and latency. Workflow error handling must be well-defined, since Spot
+instances have a significantly higher liklihood of being deallocated as compared with On-Demand
+instances.
+
+This is an example of a ComputeEnvironment that uses Spot::
```
Comment (stark525) [inline src/cirrus/docs/components/tasks/batch.rst:427]: extra `:`

File: src/cirrus/docs/components/tasks/batch.rst [hunk 5]
```diff
@@ -416,11 +416,78 @@ CPUs each job requires.
...
+
+The AllocationStrategy of `BEST_FIT_PROGRESSIVE` indicates that Spot requests should be made
+progressively for the instance types that ECS determines will best meet the resource needs, but
+that any of these machines is acceptable. This is different than the `BEST_FIT` allocation
+strategy that will pick the best one and wait until that best one can be fulfilled.
+
+These specific `InstanceTypes` were chosen because they match the processor architecture needed
+by the Docker image that will run on them (AMD64) and they most closely "pack" with the amount
+vCPU and memory configured for the container, in this case, 1 vCPU and 9.8GB memory, with their
+1 vCPU to 8GB memory ratios.
```
Comment (stark525) [inline src/cirrus/docs/components/tasks/batch.rst:485]: If you choose an AllocationStrategy with a wide range of available instance types, larger instances will be running more jobs, and you need to ensure you have appropriate storage for these large instances. For example, if each task requires 3GB of disk space, you may be able to allocate an 8x sized instance to run 8 tasks, but would need 24GB storage. This would be a waste for smaller instance types running 1 task, where the storage is over-allocated.

File: src/cirrus/docs/components/tasks/index.rst [hunk 6]
```diff
@@ -193,14 +195,89 @@ Lambda and Batch
...
 This command does the same as both of the above ``create`` command examples, so
 the listed caveats of both apply here: ensure the handler code is completed,
 and the batch configuration is updated to match the task requirements.
 
+* :doc:`Conditionally Using Batch or Lambda <../workflows/batch>`
+
+Docker Image
+---------------
+
+Unfortuantely, the same Docker image cannot be used by both Batch and Lambda, as they
```
Comment (stark525) [inline src/cirrus/docs/components/tasks/index.rst:211]: I thought we had a way for tasks to `RunBatchAsLambda` and `RunLambdaAsBatch` but I can't seem to find it anymore.

File: src/cirrus/docs/components/workflows/batch.rst [hunk 7]
```diff
@@ -1,10 +1,10 @@
 Batch tasks in workflows
```
Comment (stark525) [inline src/cirrus/docs/components/workflows/batch.rst:1]: A large section that would be great to have is Bring Your Own Container (BYOC) - "So you have a docker container, how do you modify it to work with cirrus?"

I think we've accepted that this is not going to be a first-class citizen since it varies wildly, but general guidance here would be very helpful, internally and externally.

File: src/cirrus/docs/components/tasks/index.rst (hunk 6 above)
Comment (philvarner) [reply src/cirrus/docs/components/tasks/index.rst:211]: @matthewhanson updated one the examples to use a Choice, and that attempted to use the same image, which didn't work. Matt, do you know of what Dan's referencing?

File: src/cirrus/docs/components/workflows/definitions.rst [hunk 8]
```diff
@@ -147,6 +147,49 @@ workflows. Cirrus allows several means of building such multi-stage pipelines:
   conditions don't quite fit with the in-the-box approaches.
 
 
+Error handling
+--------------
+
```
Comment (stark525) [inline src/cirrus/docs/components/workflows/definitions.rst:152]: Random thought - another thing that is confusing is the pre- and post-batch steps surrounding batch tasks in a workflow. Somewhere, we should explain why they exist, but also that a batch task may be marked as successful but fail in the post-batch step.

Caveat: I'm not sure if this is because the projects I've been on were "doing error handling wrong", often it was BYOC with a wide variety of logging and error handling strategies that we were trying to black box.

File: src/cirrus/docs/components/tasks/batch.rst (hunk 5 above)
Comment (philvarner) [reply src/cirrus/docs/components/tasks/batch.rst:485]: If you choose an AllocationStrategy with a wide range of available instance types,
larger instances will be running more jobs, and you need to ensure you have
appropriate storage for these larger instances. Unfortunately, only one value can be
set for storage, which every EC2 instance gets, no matter how many concurrent tasks it
can support. For example, if each task requires 1vCPU and 3GB of storage and you allow
r4.xlarge (4 vCPUs) to r4.24xlarge (96 vCPUs), you must specify the the maximum amount of
storage required (96 x 3GB = 288GB), even though the r4.xlarge instances will always have
276GB of unused storage that incurs cost.

File: src/cirrus/docs/components/workflows/batch.rst (hunk 7 above)
Comment (philvarner) [reply src/cirrus/docs/components/workflows/batch.rst:1]: this is a good point. some of this is described in the `Docker Image` section in Tasks, but I don't know if we define anywhere what the task.py should look like

File: src/cirrus/docs/components/tasks/index.rst (hunk 6 above)
Comment (matthewhanson) [reply src/cirrus/docs/components/tasks/index.rst:211]: There used to be a Docker image used in a generic Batch task and that could be used to run any Lambda that was deployed as a standard Lambda package (ie zip file).   The task would take in the ARN of the Lambda, fetch the deploy package code, unzip, then run the included task.py file.   It was a way to just deploy a Lambda, and not have to deploy anything else.

The utility of this isn't really there anymore.  Most importantly, we now deploy Lambdas as images to begin with, so this same image should be runnable in either Batch or Lambda. (though right now there's an issue I think with how ENTRYPOINT is set that makes these images incompatible.

Also, while the original idea of this was useful for development, in production you end up fetching the deploy package over and over again, every time you run a job.

tl;dr - now with stac-task you develop and deploy a single image, which can then be used as a Lambda or Batch task.

File: src/cirrus/docs/components/workflows/definitions.rst (hunk 8 above)
Comment (philvarner) [reply src/cirrus/docs/components/workflows/definitions.rst:152]: https://cirrus-geo.github.io/cirrus-geo/stable/cirrus/components/workflows/batch.html#pre-batch-and-post-batch ?

File: src/cirrus/docs/components/workflows/batch.rst (hunk 7 above)
Comment (philvarner) [reply src/cirrus/docs/components/workflows/batch.rst:1]: added

**Review by stark525 — APPROVED** [review]
